sdk/program/src/message/versions.rs

```diff
@@ -0,0 +1,262 @@
+use {
+    crate::{
+        hash::Hash,
+        instruction::CompiledInstruction,
+        message::{v0, Message, MessageHeader},
+        pubkey::Pubkey,
+        sanitize::{Sanitize, SanitizeError},
+        short_vec,
+    },
+    serde::{
+        de::{self, Deserializer, SeqAccess, Visitor},
+        ser::{SerializeTuple, Serializer},
+        {Deserialize, Serialize},
+    },
+    std::fmt,
+};
+
+/// Bit mask that indicates whether a serialized message is versioned.
+const VERSION_PREFIX: u8 = 0x80;
+
+/// Message versions supported by the Solana runtime.
+///
+/// # Serialization
+///
+/// If the first bit is set, the remaining 7 bits will be used to determine
+/// which message version is serialized starting from version `0`. If the first
+/// is bit is not set, all bytes are used to encode the original `Message`
+/// format.
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub enum MessageVersions {
```

Comment:
if you find it hard to place `frozen_abi` because of this byte-clamping versioning scheme for transactions. let me know. I can tweak frozen_abi to work, if possible.

Reply:
I forgot to add it! It works great, thanks :)